lib/handler/pool.ex

```diff
@@ -21,19 +35,27 @@ defmodule Handler.Pool do
     {:ok, state}
   end
 
-  def handle_call({:run, fun, opts}, from, state) do
-    case State.start_worker(state, fun, opts, from) do
-      {:ok, state} ->
-        {:noreply, state}
+  def handle_call({:run, fun, opts}, {pid, _tag}, state) do
+    case State.start_worker(state, fun, opts, pid) do
+      {:ok, state, ref} ->
```

Comment:
One of the big changes here is that running a function in the pool actually gets back 2 messages now, the first one is an acceptance message that provide a unique `reference` for the job and the second one has the results of the function call.

This means slightly more overhead, but it gives us the ability to support an asynchronous API in addition to the synchronous API. It also means that if the pool itself ever becomes unresponsive (backed up with messages) we'll get specific errors around the `GenServer.call`.